Search text highlighting

codemonkey800 opened the pull request:
## Description

This PR implements text highlighting for search results. This works by:

1. Getting match start and end indices using [includeMatches](https://fusejs.io/api/options.html#includematches) option in Fuse.js
2. Finding the most similar match using [string-similarity](https://www.npmjs.com/package/string-similarity)
3. Highlighting the text using [react-highlight-words](https://github.com/bvaughn/react-highlight-words) 

### Demo

Demo Link: https://napari-hub-search-highlighting-demo.vercel.app

#### Search Resulting Highlighting

![image](https://user-images.githubusercontent.com/2176050/120540231-4df33b80-c39d-11eb-9a74-9ac5a9291e11.png)

liaprins-czi reviewed (approved):
This is great!

justinelarsen reviewed (approved):
LGTM, this is a really cool feature! Nice work @codemonkey800 ! 

General design comment: The number of characters in the description if it's highlighted is too small in my opinion. It's difficult to know any context from just the small amount. I would recommend at least double the amount of characters to get some relevant context. @liaprins-czi What do you think?

liaprins-czi commented:
Yes, good point @Justine Larsen ***@***.***> ! I was
imagining it'd be ideal if the length of snippet is approximately as long
as the width of the two columns in the widest screen (~600px) I think. So
for number of characters it looks like doubling it would be perfect and
help the context a lot more

On Wed, Jun 2, 2021 at 3:16 PM Justine Larsen ***@***.***>
wrote:

> ***@***.**** approved this pull request.
>
> LGTM, this is a really cool feature! Nice work @codemonkey800
> <https://github.com/codemonkey800> !
>
> General design comment: The number of characters in the description if
> it's highlighted is too small in my opinion. It's difficult to know any
> context from just the small amount. I would recommend at least double the
> amount of characters to get some relevant context. @liaprins-czi
> <https://github.com/liaprins-czi> What do you think?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/chanzuckerberg/napari-hub/pull/65#pullrequestreview-674726230>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AMWPDSESQIXYXJKG5DOSWDTTQ2UTBANCNFSM457KT4RA>
> .
>


-- 

*Lia Prins*

Senior Data Visualization Designer

(650) 422-1750

chanzuckerberg.com

kne42 reviewed (approved):
Woah cool! :o LGTM :)

codemonkey800 commented:
1. I updated the description preview to be 40 characters on each side now. I also refactored the function a bit to handle edge cases where the matched term shows up at the beginning or end of the description.
2. I refactored matching finding to use [string-similarity](https://www.npmjs.com/package/string-similarity) instead of the first match. This gives us way better results than the previous method. I also did a Lighthouse comparison and found the performance difference negligible (almost zero difference).
